src/mixins/createI18nMixin.ts

```diff
@@ -0,0 +1,362 @@
+ * with the state being updated again when the correct locale dictionary has been loaded.
+ */
+function setLabelState(instance: I18nMixin<Messages, I18nState>, labels: I18nLabels, messages?: Messages) {
+	const bundles = bundleMap.get(instance);
+	const bundlePaths = Object.keys(bundles);
+	const bundleCount = bundlePaths.length;
+	let updateState = false;
+
+	const state = Object.keys(labels).reduce((state, key) => {
+		let [ bundlePath, labelKey ] = labels[key].split(':');
+		let localeMessages = messages;
+
+		// If there is only one registered bundle, assume there is no bundle path.
+		if (bundleCount === 1 && !labelKey) {
+			labelKey = bundlePath;
+			bundlePath = '';
+		}
+		else if (!bundlePath || !labelKey) {
+			throw new Error('I18n labels must be specified in the format `${bundlePath}:${key}`.');
+		}
+
+		if (!localeMessages) {
+			if (!bundlePath) {
+				bundlePath = bundlePaths[0];
+			}
+
+			const bundle = bundles[bundlePath];
+			localeMessages = bundle && instance.localizeBundle(bundle);
+		}
+
+		if (localeMessages) {
+			if (typeof localeMessages[labelKey] !== 'string') {
+				throw new Error(`${bundlePath} has no "${labelKey}" key.`);
+			}
+
+			updateState = true;
+			state[key] = localeMessages[labelKey];
+		}
+
+		return state;
+	}, {} as I18nState);
+
+	if (updateState) {
+		instance.setState(state);
+	}
+}
+
+/**
+ * @private
+ * Update the instance's state if `state.locale` or `state.labels` has changed.
+ */
+function updateLabels(this: I18nMixin<Messages, I18nState>) {
+	const { labels, locale } = this.state;
+	const needsUpdate = localeChanged(localeMap.get(this), locale) || labelsChanged(labelMap.get(this), labels);
+
+	if (labels && needsUpdate) {
+		labelMap.set(this, labels);
+		localeMap.set(this, locale);
+		setLabelState(this, labels);
+	}
+}
+
+const createI18nMixin: I18nFactory = compose<I18n<Messages>, I18nOptions<I18nState>>({
+	nodeAttributes: [
+		function (this: I18nMixin<Messages, I18nState>, attributes: VNodeProperties): VNodeProperties {
+			const properties = {
+				'data-locale': null,
+				dir: null
+			} as I18nVNodeProperties;
+			const { locale, rtl } = this.state;
+
+			if (typeof rtl === 'boolean') {
+				properties['dir'] = rtl ? 'rtl' : 'ltr';
+			}
+
+			if (locale) {
+				properties['data-locale'] = locale;
+			}
+
+			return properties;
+		}
+	],
+
+	localizeBundle(this: I18nMixin<Messages, I18nState>, bundle: Bundle<Messages>): Messages {
+		this.registerBundle(bundle);
+		return getLocaleMessages(this, bundle) || bundle.messages;
+	},
+
+	registerBundle(this: I18nMixin<Messages, I18nState>, bundle: Bundle<Messages>) {
+		const { bundlePath } = bundle;
+		const registered = bundleMap.get(this);
+
+		if (!registered[bundlePath]) {
+			registered[bundlePath] = bundle;
+		}
+	}
+}, (instance: I18nMixin<Messages, I18nState>, options?: I18nOptions<I18nState>) => {
+	bundleMap.set(instance, {} as BundleMap<Messages>);
+	const bundles = options && options.bundles || instance.bundles;
```

Comment:
I think we should merge them, because you could see a situation where, in a complex widget, where you would have some strings that are coming from the class and some that you might want to configure on the instance, but still have the ones that were baked into the class.